synapse/handlers/presence.py

```diff
@@ -1249,6 +1308,27 @@ async def set_state(
 
         prev_state = await self.current_state_for_user(user_id)
 
+        # Always update the device specific information.
+        device_state = self.user_to_device_to_current_state.setdefault(
+            user_id, {}
+        ).setdefault(
+            device_id,
```

Comment:
Do we want to do this if `device_id` is None? As nothing will ever overwrite it?

Reply:
I think in some unit tests, the device_id is set to be `None`. That shouldn't occur in the real/outside world though, would it?

Reply:
> As nothing will ever overwrite it?

Why would nothing ever overwrite it? There are few types of users which have access tokens without associated device IDs:

* Guest users
* Appservice users
* Access tokens created using admin APIs

Reply:
I'm not super thrilled to have to support `None` in all of this code though I also ended up using it for the special pseudo-device created on-start.

Reply:
Let's continue this on https://github.com/matrix-org/synapse/pull/16171/files#r1303245932.